File: problems/147_insertion-sort-list/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        if (!head || !head->next)
            return head;
        ListNode dummy(0);
        ListNode *curr = head;
        while (curr)
        {
            ListNode *nextNode = curr->next;
            ListNode *prev = &dummy;
            while (prev->next && prev->next->val < curr->val)
                prev = prev->next;
            curr->next = prev->next;
            prev->next = curr;
            curr = nextNode;
        }
        return dummy.next;
    }
};

ListNode *build_list(const vector<int> &values)
{
    ListNode dummy(0);
    ListNode *tail = &dummy;
    for (int value : values)
    {
        tail->next = new ListNode(value);
        tail = tail->next;
    }
    return dummy.next;
}

vector<int> list_to_vector(ListNode *head)
{
    vector<int> out;
    while (head)
    {
        out.push_back(head->val);
        head = head->next;
    }
    return out;
}

string vec_to_string(const vector<int> &nums)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < nums.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << nums[i];
    }
    ss << "]";
    return ss.str();
}

void free_list(ListNode *head)
{
    while (head)
    {
        ListNode *next = head->next;
        delete head;
        head = next;
    }
}

int main()
{
    Solution sol;
    vector<vector<int>> samples = {
        {4, 2, 1, 3},
        {-1, 5, 3, 4, 0},
    };
    vector<vector<int>> expected = {
        {1, 2, 3, 4},
        {-1, 0, 3, 4, 5},
    };

    for (size_t i = 0; i < samples.size(); ++i)
    {
        ListNode *head = build_list(samples[i]);
        ListNode *result_head = sol.insertionSortList(head);
        vector<int> result = list_to_vector(result_head);

        cout << "Sample " << (i + 1) << ": head = " << vec_to_string(samples[i]) << "\n";
        cout << "Placeholder result: " << vec_to_string(result) << "\n";
        cout << "Expected output: " << vec_to_string(expected[i]) << "\n\n";

        free_list(result_head);
    }

    return 0;
}

File: problems/147_insertion-sort-list/main.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    pub fn insertion_sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() || head.as_ref().unwrap().next.is_none() {
            return head;
        }
        let mut dummy = Box::new(ListNode::new(0));
        let mut curr = head;
        while let Some(mut node) = curr {
            curr = node.next.take();
            let mut prev = &mut dummy;
            while prev.next.is_some() && prev.next.as_ref().unwrap().val < node.val {
                prev = prev.next.as_mut().unwrap();
            }
            node.next = prev.next.take();
            prev.next = Some(node);
        }
        dummy.next
    }
}

fn build_list(values: &[i32]) -> Option<Box<ListNode>> {
    let mut head: Option<Box<ListNode>> = None;
    for &value in values.iter().rev() {
        let mut node = Box::new(ListNode::new(value));
        node.next = head;
        head = Some(node);
    }
    head
}

fn list_to_vec(mut head: Option<Box<ListNode>>) -> Vec<i32> {
    let mut out = Vec::new();
    while let Some(node) = head {
        out.push(node.val);
        head = node.next;
    }
    out
}

fn main() {
    let samples = vec![
        (vec![4, 2, 1, 3], vec![1, 2, 3, 4]),
        (vec![-1, 5, 3, 4, 0], vec![-1, 0, 3, 4, 5]),
    ];

    for (idx, (input, expected)) in samples.into_iter().enumerate() {
        let head = build_list(&input);
        let result_head = Solution::insertion_sort_list(head);
        let result = list_to_vec(result_head);

        println!("Sample {}: head = {:?}", idx + 1, input);
        println!("Placeholder result: {:?}", result);
        println!("Expected output: {:?}\n", expected);
    }
}
